control_plane/src/postgresql_conf.rs

```diff
@@ -0,0 +1,78 @@
+///
+/// Module for parsing postgresql.conf file.
+///
+/// NOTE: This doesn't implement the full, correct postgresql.conf syntax. Just
+/// enough to extract a few settings we need in Zenith, assuming you don't do
+/// funny stuff like include-directives or funny escaping.
+use anyhow::Result;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::collections::HashMap;
+use std::io::BufRead;
+
+
+/// In-memory representation of a postgresql.conf file
+pub struct PostgresConf {
+    lines: Vec<String>,
+    hash: HashMap<String, String>,
+}
+
+lazy_static! {
+    static ref CONF_LINE_RE: Regex = Regex::new(r"^((?:\w|\.)+)\s*=\s*(\S+)$").unwrap();
+}
+
+/// Unescape a possibly-quoted value.
+///
+/// FIXME: This doesn't handle '' and backslashes within the string currently.
+/// Only really removes quotes at the be beginning and end of the string.
+fn unescape_str<'a>(s: &'a str) -> &'a str {
+    // If the string has a quote at the beginning and end, strip them out.
+    //
+    if s.len() >= 2 && s.starts_with('\'') && s.ends_with('\'') {
+        &s[1..(s.len() - 1)]
+    } else {
+        s
+    }
+}
+
+impl PostgresConf {
+
+    /// Read file into memory
+    pub fn read<R>(read: &mut R) -> Result<PostgresConf>
```

Comment:
Thanks, will switch to `&mut impl std::io::Read`, that's simpler. I wasn't aware of that syntax. Does it mean the same thing, or is there some subtle difference?

Reply:
Semantically, it's the same thing. The only practical difference is that you can't specify the generic argument manually. I've run into that in the past, but it's typically not too much of an issue (IIRC it's always a non-breaking change to switch to explicit generics).

Also! You should be able to just use `impl std::io::Read` here, without the leading `&mut` because there's a blanket implementation of `io::Read` for `&mut R where R: io::Read`.

Reply:
> Semantically, it's the same thing. The only practical difference is that you can't specify the generic argument manually. I've run into that in the past, but it's typically not too much of an issue (IIRC it's always a non-breaking change to switch to explicit generics).

Gotcha.

> Also! You should be able to just use `impl std::io::Read` here, without the leading `&mut` because there's a blanket implementation of `io::Read` for `&mut R where R: io::Read`.

Oh, ok. I don't really understand what's happening here, but you're right, it works :-).